README
NAME

    FTN::Crypt - Encryption of the FTN messages.

 VERSION

    0.5

SYNOPSIS

        use FTN::Crypt;
    
        my $obj = FTN::Crypt->new(
            Nodelist => 'NODELIST.*',
            Pointlist => [
                'pointlist_1.*',
                'pointlist_2',
            ],
        ) or die FTN::Crypt->error;
        
        $obj->encrypt_message(
            Address => $ftn_address,
            Message => $msg,
        ) or die $obj->error;

DESCRIPTION

    The possibility of FTN netmail encryption may be sometimes a useful
    option. Corresponding nodelist flag was proposed in FSC-0073.

    Although current FidoNet Policy (version 4.07 dated June 9, 1989)
    clearly forbids routing of encrypted traffic without the express
    permission of all the links in the delivery system, it's still possible
    to deliver such messages directly. And, obviously, such routing may be
    allowed in FTN networks other than FidoNet.

    The proposed nodelist userflag is ENCRYPT:[TYPE], where [TYPE] is one
    of 'PGP2', 'PGP5', 'GnuPG'. So encryption-capable node should have
    something like U,ENCRYPT:PGP5 in his nodelist record.

METHODS

 new()

    Constructor.

  Parameters:

      * Nodelist: Path to nodelist file(s), either scalar or arrayref. If
      contains wildcard, file with maximum number in digital extension will
      be selected.

      * Optional Pointlist: Path to pointlist file(s), either scalar or
      arrayref. If contains wildcard, file with maximum number in digital
      extension will be selected.

      * Optional Keyserver Keyserver (defaults to
      'https://zimmermann.mayfirst.org/pks/lookup').

      * Optional Pubring Public keyring file.

      * Optional Secring Secret keyring file.

  Returns:

    Created object or error in FTN::Crypt->error.

    Sample:

        my $obj = FTN::Crypt->new(
            Nodelist => 'NODELIST.*',
            Pointlist => [
                'pointlist_1.*',
                'pointlist_2',
            ],
        ) or die FTN::Crypt->error;

 encrypt_message()

    Message encryption.

  Parameters:

      * Address: Recipient's FTN address.

      * Message: FTN message text with kludges.

  Returns:

    Encrypted message or error in $obj->error.

    Sample:

        my $res = $obj->encrypt_message(
            Address => $ftn_address,
            Message => $msg,
        ) or die $obj->error;

 decrypt_message()

    Message decryption.

  Parameters:

      * Address: Recipient's FTN address.

      * Message: FTN message text with kludges.

      * Passphrase: Key passphrase.

  Returns:

    Decrypted message or error in $obj->error.

    Sample:

        my $res = $obj->decrypt_message(
            Address => $ftn_address,
            Message => $msg,
            Passphrase => $pass,
        ) or die $obj->error;

AUTHOR

    Petr Antonov, <petr@antonov.space>

COPYRIGHT AND LICENSE

    Copyright (C) 2019 by Petr Antonov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl 5.10.0. For more details, see the full
    text of the licenses at https://opensource.org/licenses/Artistic-1.0,
    and http://www.gnu.org/licenses/gpl-2.0.html.

    This package is provided "as is" and without any express or implied
    warranties, including, without limitation, the implied warranties of
    merchantability and fitness for a particular purpose.

INSTALLATION

    Using cpan:

        $ cpan FTN::Crypt

    Manual install:

        $ perl Makefile.PL
        $ make
        $ make test
        $ make install

REFERENCES

    1 FidoNet Policy Document Version 4.07
    <https://www.fidonet.org/policy4.txt>

    2 FTS-5001 - Nodelist flags and userflags
    <http://ftsc.org/docs/fts-5001.006>

    3 FSC-0073 - Encrypted message identification for FidoNet *Draft I*
    <http://ftsc.org/docs/fsc-0073.001>



Changes
Revision history for FTN::Crypt

0.5     2019-11-18
    [*] Add pointlist processing.
    [*] Multiple node- and pointlists.

0.4     2019-11-17
    [*] Rewrite message processing code.

0.3.0   2019-11-15
    [*] Add tests and prepare CPAN release.

0.2.5   2019-11-01
    [*] Switch to Gnupg::Interface, pure-perl OpenPGP implementation in
        Crypt::OpenPGP seems to be broken.
    [*] New error handling.

0.1.0   2019-10-29
    [*] Initial release.



Makefile.PL
use v5.10.1;
use strict;
use warnings;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
	NAME             => 'FTN::Crypt',
	VERSION_FROM     => 'lib/FTN/Crypt.pm',
	ABSTRACT_FROM    => 'lib/FTN/Crypt.pm',
	AUTHOR           => 'Petr Antonov',
	LICENSE          => 'perl',
	MIN_PERL_VERSION => '5.010',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    TEST_REQUIRES => {
        'Pod::Coverage'       => '0',
		'Test::CheckManifest' => '0',
		'Test::Data'          => '0',
		'Test::LongString'    => '0',
        'Test::More'          => '0',
        'Test::Pod'           => '0',
        'Test::Pod::Coverage' => '0',
    },
	PREREQ_PM        => {
		'strict'           => '0',
		'warnings'         => '0',
		'Try::Tiny'        => '0.28',
		'IO::Handle'       => '1.35',
		'FTN::Address'     => '1.04',
		'FTN::Nodelist'    => '1.08',
		'GnuPG::Interface' => '0.52',
		'PGP::Finger'      => '1.1',
	},
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'FTN-Crypt-*' },
	(eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
		  'meta-spec' => { version => 2 },
		  resources => {
			  repository => {
				  type => 'git',
				  url  => 'https://github.com/pietro-a/FTN-Crypt.git',
				  web  => 'https://github.com/pietro-a/FTN-Crypt',
			  },
		  }})
		: ()
	),
);

# Compatibility with old versions of ExtUtils::MakeMaker
unless (eval { ExtUtils::MakeMaker->VERSION('6.64'); 1 }) {
    my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES} || {};
    @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

unless (eval { ExtUtils::MakeMaker->VERSION('6.55_03'); 1 }) {
    my $build_requires = delete $WriteMakefileArgs{BUILD_REQUIRES} || {};
    @{$WriteMakefileArgs{PREREQ_PM}}{keys %$build_requires} = values %$build_requires;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
    unless eval { ExtUtils::MakeMaker->VERSION('6.52'); 1 };
delete $WriteMakefileArgs{MIN_PERL_VERSION}
    unless eval { ExtUtils::MakeMaker->VERSION('6.48'); 1 };
delete $WriteMakefileArgs{LICENSE}
    unless eval { ExtUtils::MakeMaker->VERSION('6.31'); 1 };

WriteMakefile(%WriteMakefileArgs);


lib/FTN/Crypt/Msg.pm
# FTN::Crypt::Msg - Message parsing for the FTN::Crypt module
#
# Copyright (C) 2019 by Petr Antonov
#
# This library is free software; you can redistribute it and/or modify it
# under the same terms as Perl 5.10.0. For more details, see the full text
# of the licenses at https://opensource.org/licenses/Artistic-1.0, and
# http://www.gnu.org/licenses/gpl-2.0.html.
#
# This package is provided "as is" and without any express or implied
# warranties, including, without limitation, the implied warranties of
# merchantability and fitness for a particular purpose.
#
package FTN::Crypt::Msg::MsgChunk;

use strict;
use warnings;
use v5.10.1;

sub new {
    my $class = shift;

    my $self = {
        data => [],
    };
    
    bless $self, $class;
}

sub get {
    my $self = shift;
    
    return $self->{data};
}

sub set {
    my $self = shift;
    my ($line) = @_;
    
    push @{$self->{data}}, $line;
}

#----------------------------------------------------------------------#

package FTN::Crypt::Msg::KludgeChunk;

use strict;
use warnings;
use v5.10.1;

use base qw/FTN::Crypt::Msg::MsgChunk/;

sub remove {
    my $self = shift;
    my ($kludge) = @_;
    
    if (defined $kludge && $kludge ne '') {
        @{$self->{data}} = grep { !/^${kludge}(?::?\s.+)*$/ }
                           @{$self->{data}};
    } else {
        return;
    }

    return 1;
}

#----------------------------------------------------------------------#

package FTN::Crypt::Msg::TextChunk;

use strict;
use warnings;
use v5.10.1;

use base qw/FTN::Crypt::Msg::MsgChunk/;

sub get {
    my $self = shift;
    my ($ftn_ready) = @_;

    my $sep = $ftn_ready ? "\r" : "\n";

    return join $sep, @{$self->{data}};
}

#----------------------------------------------------------------------#

package FTN::Crypt::Msg;

use strict;
use warnings;
use v5.10.1;

use base qw/FTN::Crypt::Error/;

#----------------------------------------------------------------------#

=head1 NAME

FTN::Crypt::Msg - Message parsing for the L<FTN::Crypt> module.

=head1 SYNOPSIS

    use FTN::Crypt::Msg;

    my $obj = FTN::Crypt::Msg->new(
        Address => $ftn_address,
        Message => $msg,
    );
    $obj->add_kludge('ENC: PGP5');
    $obj->remove_kludge('ENC');
    my $text = $obj->get_text;
    my $kludges = $obj->get_kludges;
    my $msg = $obj->get_message;

=cut

#----------------------------------------------------------------------#

use FTN::Address;

#----------------------------------------------------------------------#

my $SOH = chr(1);

my %PREDEFINED_INDEX = (
    'TOP'    => 0,
    'BOTTOM' => -1,
);
my %DEFAULT_INDEX = (
    KLUDGE => 'TOP',
    TEXT   => 'BOTTOM',
);

#----------------------------------------------------------------------#

=head1 METHODS

=cut

#----------------------------------------------------------------------#

=head2 new()

Constructor.

=head3 Parameters:

=over 4

=item * C<Address>: Recipient's FTN address.

=item * C<Message>: FTN message text with kludges.

=back

=head3 Returns:

Created object or error in C<FTN::Crypt::Msg-E<gt>error>.

Sample:

    my $obj = FTN::Crypt::Msg->new(
        Address => $ftn_address,
        Message => $msg,
    ) or die FTN::Crypt::Msg->error;

=cut

sub new {
    my $class = shift;
    my (%opts) = @_;

    unless (%opts) {
        $class->set_error('No options specified');
        return;
    }
    unless ($opts{Address}) {
        $class->set_error('No address specified');
        return;
    }
    unless ($opts{Message}) {
        $class->set_error('No message specified');
        return;
    }

    my $self = {
        msg => [],
        idx => {
            KLUDGE => [],
            TEXT   => [],
        },
    };

    $self = bless $self, $class;

    unless ($self->set_address($opts{Address})) {
        $class->set_error($self->error);
        return;
    }
    unless ($self->set_message($opts{Message})) {
        $class->set_error($self->error);
        return;
    }

    return $self;
}

#----------------------------------------------------------------------#

sub _check_kludge {
    my $self = shift;
    my ($kludge) = @_;

    unless (defined $kludge && $kludge ne '') {
        $self->set_error('Kludge is empty');
        return;
    }

    return $kludge;
}

#----------------------------------------------------------------------#

sub _check_text {
    my $self = shift;
    my ($text) = @_;

    unless (defined $text && $text ne '') {
        $self->set_error('Text is empty');
        return;
    }

    return $text;
}

#----------------------------------------------------------------------#

sub _check_idx {
    my $self = shift;
    my ($type, $idx) = @_;

    unless (grep /^$type$/, keys %{$self->{idx}}) {
        $self->set_error("Invalid message area type (`$type')");
        return;        
    }

    $idx = $DEFAULT_INDEX{$type} unless defined $idx;

    $idx = $PREDEFINED_INDEX{$idx} if defined $PREDEFINED_INDEX{$idx};
    
    unless ($idx =~ /^-?\d+$/) {
        $self->set_error("Invalid chunk index (`$idx')");
        return;
    }

    unless (defined $self->{idx}->{$type}->[$idx]) {
        $self->set_error("Invalid chunk index (`$idx')");
        return;
    }

    return $idx;
}

#----------------------------------------------------------------------#

=head2 add_kludge()

Add kludge to the message.

=head3 Parameters:

=over 4

=item * Kludge string.

=item * B<Optional> C<[TOP|BOTTOM|<index>]> Kludges block, defaults to TOP.

=back

=head3 Returns:

True or error in C<$obj-E<gt>error>.

Sample:

    $obj->add_kludge('ENC: PGP5') or die $obj->error;

=cut

sub add_kludge {
    my $self = shift;
    my ($kludge, $idx) = @_;

    $kludge = $self->_check_kludge($kludge);
    $idx = $self->_check_idx('KLUDGE', $idx);

    if (defined $kludge && defined $idx) {
        $self->{msg}->[$self->{idx}->{KLUDGE}->[$idx]]->set($kludge);
    } else {
        return;
    }

    return 1;
}

#----------------------------------------------------------------------#

=head2 remove_kludge()

Remove kludge from the message.

=head3 Parameters:

=over 4

=item * Kludge string, may be only the first part of the composite kludge.

=item * B<Optional> C<[TOP|BOTTOM|<index>]> Kludges block, defaults to TOP.

=back

=head3 Returns:

True or error in C<$obj-E<gt>error>.

Sample:

    $obj->remove_kludge('ENC') or die $obj->error;

=cut

sub remove_kludge {
    my $self = shift;
    my ($kludge, $idx) = @_;

    $kludge = $self->_check_kludge($kludge);
    $idx = $self->_check_idx('KLUDGE', $idx);

    if (defined $kludge && defined $idx) {
        $self->{msg}->[$self->{idx}->{KLUDGE}->[$idx]]->remove($kludge);
    } else {
        return;
    }

    return 1;
}

#----------------------------------------------------------------------#

=head2 get_kludges()

Get message kludges.

=head3 Parameters:

None.

=head3 Returns:

Arrayref with kludges list or error in C<$obj-E<gt>error>.

Sample:

    $obj->get_kludges() or die $obj->error;

=cut

sub get_kludges {
    my $self = shift;

    my $kludges = [];
    foreach my $c (@{$self->{msg}}) {
        push @{$kludges}, $c->get if $c->isa('FTN::Crypt::Msg::KludgeChunk');
    }

    return $kludges;
}

#----------------------------------------------------------------------#

=head2 get_address()

Get recipient's FTN address.

=head3 Parameters:

None.

=head3 Returns:

Recipient's FTN address or error in C<$obj-E<gt>error>.

Sample:

    my $ftn_address = $obj->get_address() or die $obj->error;

=cut

sub get_address {
    my $self = shift;

    my $addr = $self->{addr}->get;
    unless ($addr) {
        $self->set_error($@);
        return;
    }

    return $addr;
}

#----------------------------------------------------------------------#

=head2 set_address()

Set recipient's FTN address.

=head3 Parameters:

=over 4

=item * Recipient's FTN address.

=back

=head3 Returns:

True or error in C<$obj-E<gt>error>.

Sample:

    $obj->set_address($ftn_address)

=cut

sub set_address {
    my $self = shift;
    my ($addr) = @_;

    $self->{addr} = FTN::Address->new($addr);
    unless ($self->{addr}) {
        $self->set_error($@);
        return;
    }

    return 1;
}

#----------------------------------------------------------------------#

=head2 get_text()

Get text part of the message.

=head3 Parameters:

=over 4

=item * B<Optional> C<[TOP|BOTTOM|<index>]> Text block, defaults to BOTTOM.

=back

=head3 Returns:

Text part of the message or error in C<$obj-E<gt>error>.

Sample:

    my $text = $obj->get_text() or die $obj->error;

=cut

sub get_text {
    my $self = shift;
    my ($idx) = @_;

    $idx = $self->_check_idx('TEXT', $idx);

    my $text = '';
    if (defined $idx) {
        $text = $self->{msg}->[$self->{idx}->{TEXT}->[$idx]]->get;
    } else {
        return;
    }

    return $text;
}

#----------------------------------------------------------------------#

=head2 get_all_text()

Get all text parts of the message.

=head3 Parameters:

None.

=head3 Returns:

Arrayref with text parts of the message or error in C<$obj-E<gt>error>.

Sample:

    my $text = $obj->get_all_text() or die $obj->error;

=cut

sub get_all_text {
    my $self = shift;

    my $text = [];
    foreach my $c (@{$self->{msg}}) {
        push @{$text}, $c->get if $c->isa('FTN::Crypt::Msg::TextChunk');
    }

    return $text;
}

#----------------------------------------------------------------------#

=head2 set_text()

Set text part of the message.

=head3 Parameters:

=over 4

=item * Text part of the message.

=item * B<Optional> C<[TOP|BOTTOM|<index>]> Text block, defaults to BOTTOM.

=back

=head3 Returns:

True or error in C<$obj-E<gt>error>.

Sample:

    $obj->set_text($text) or die $obj->error;

=cut

sub set_text {
    my $self = shift;
    my ($text, $idx) = @_;

    $text = $self->_check_text($text);
    $idx = $self->_check_idx('TEXT', $idx);

    if (defined $text && defined $idx) {
        $self->{msg}->[$self->{idx}->{TEXT}->[$idx]] = FTN::Crypt::Msg::TextChunk->new;
        $text =~ s/\r\n/\r/g;
        $text =~ s/\n/\r/g;
        my @text_lines = split /\r/, $text;
        foreach my $l (@text_lines) {
            $self->{msg}->[$self->{idx}->{TEXT}->[$idx]]->set($l);
        }
    } else {
        return;
    }

    return 1;
}

#----------------------------------------------------------------------#

=head2 get_message()

Get FTN message text with kludges.

=head3 Parameters:

None.

=head3 Returns:

FTN message text with kludges or error in C<$obj-E<gt>error>.

Sample:

    my $msg = $obj->get_message() or die $obj->error;

=cut

sub get_message {
    my $self = shift;

    my @msg;
    foreach my $c (@{$self->{msg}}) {
        if ($c->isa('FTN::Crypt::Msg::KludgeChunk')) {
            push @msg, join "\r", map { "${SOH}$_" } @{$c->get};
        } elsif ($c->isa('FTN::Crypt::Msg::TextChunk')) {
            push @msg, $c->get(1);
        }
    }

    my $msg_out = join "\r", @msg;
    
    return $msg_out;
}

#----------------------------------------------------------------------#

=head2 set_message()

Set FTN message text with kludges.

=head3 Parameters:

=over 4

=item * FTN message text with kludges.

=back

=head3 Returns:

True or error in C<$obj-E<gt>error>.

Sample:

    $obj->set_message($msg) or die $obj->error;

=cut

sub set_message {
    my $self = shift;
    my ($msg) = @_;

    $self->{msg} = [];
    foreach my $a (keys %{$self->{idx}}) {
        $self->{idx}->{$a} = [];
    }

    $msg =~ s/\r\n/\r/g;
    $msg =~ s/\n/\r/g;
    my @msg_lines = split /\r/, $msg;

    my $is_kludge;
    foreach my $l (@msg_lines) {
        if ($l =~ s/^${SOH}//) {
            if (!defined $is_kludge || !$is_kludge) {
                push @{$self->{msg}}, FTN::Crypt::Msg::KludgeChunk->new;
                push @{$self->{idx}->{KLUDGE}}, $#{$self->{msg}};
                $is_kludge = 1;
            }
        } else {
            if (!defined $is_kludge || $is_kludge) {
                push @{$self->{msg}}, FTN::Crypt::Msg::TextChunk->new;
                push @{$self->{idx}->{TEXT}}, $#{$self->{msg}};
                $is_kludge = 0;
            }
        }
        $self->{msg}->[-1]->set($l);
    }

    return 1;
}

1;
__END__

=head1 AUTHOR

Petr Antonov, E<lt>petr _at_ antonov _dot_ spaceE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2019 by Petr Antonov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses at L<https://opensource.org/licenses/Artistic-1.0>, and
L<http://www.gnu.org/licenses/gpl-2.0.html>.

This package is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantability and fitness for a particular purpose.

=cut


lib/FTN/Crypt.pm
# FTN::Crypt - Encryption of the FTN messages
#
# Copyright (C) 2019 by Petr Antonov
#
# This library is free software; you can redistribute it and/or modify it
# under the same terms as Perl 5.10.0. For more details, see the full text
# of the licenses at https://opensource.org/licenses/Artistic-1.0, and
# http://www.gnu.org/licenses/gpl-2.0.html.
#
# This package is provided "as is" and without any express or implied
# warranties, including, without limitation, the implied warranties of
# merchantability and fitness for a particular purpose.
#

package FTN::Crypt;

use strict;
use warnings;
use v5.10.1;

use base qw/FTN::Crypt::Error/;

#----------------------------------------------------------------------#

=head1 NAME

FTN::Crypt - Encryption of the FTN messages.

=head2 VERSION

0.5

=cut

our $VERSION = '0.5';

#----------------------------------------------------------------------#

=head1 SYNOPSIS

    use FTN::Crypt;

    my $obj = FTN::Crypt->new(
        Nodelist => 'NODELIST.*',
        Pointlist => [
            'pointlist_1.*',
            'pointlist_2',
        ],
    ) or die FTN::Crypt->error;
    
    $obj->encrypt_message(
        Address => $ftn_address,
        Message => $msg,
    ) or die $obj->error;

=head1 DESCRIPTION

The possibility of FTN netmail encryption may be sometimes a useful option.
Corresponding nodelist flag was proposed in FSC-0073.

Although current FidoNet Policy (version 4.07 dated June 9, 1989) clearly
forbids routing of encrypted traffic without the express permission of
all the links in the delivery system, it's still possible to deliver such
messages directly. And, obviously, such routing may be allowed in FTN
networks other than FidoNet.

The proposed nodelist userflag is ENCRYPT:[TYPE], where [TYPE] is one of
'PGP2', 'PGP5', 'GnuPG'. So encryption-capable node should have something
like U,ENCRYPT:PGP5 in his nodelist record.

=cut

#----------------------------------------------------------------------#

use FTN::Crypt::Constants;
use FTN::Crypt::Msg;
use FTN::Crypt::Nodelist;

use GnuPG::Interface;

use IO::Handle;

use PGP::Finger;

use Try::Tiny;

#----------------------------------------------------------------------#

my $DEFAULT_KEYSERVER_URL = 'https://zimmermann.mayfirst.org/pks/lookup';

my $GPG2_BVER = '2.1.0';

#----------------------------------------------------------------------#

=head1 METHODS

=cut

#----------------------------------------------------------------------#

=head2 new()

Constructor.

=head3 Parameters:

=over 4

=item * C<Nodelist>: Path to nodelist file(s), either scalar or arrayref. If contains wildcard, file with maximum number in digital extension will be selected.

=item * B<Optional> C<Pointlist>: Path to pointlist file(s), either scalar or arrayref. If contains wildcard, file with maximum number in digital extension will be selected.

=item * B<Optional> C<Keyserver> Keyserver (defaults to 'https://zimmermann.mayfirst.org/pks/lookup').

=item * B<Optional> C<Pubring> Public keyring file.

=item * B<Optional> C<Secring> Secret keyring file.

=back

=head3 Returns:

Created object or error in C<FTN::Crypt-E<gt>error>.

Sample:

    my $obj = FTN::Crypt->new(
        Nodelist => 'NODELIST.*',
        Pointlist => [
            'pointlist_1.*',
            'pointlist_2',
        ],
    ) or die FTN::Crypt->error;

=cut

sub new {
    my $class = shift;
    my (%opts) = @_;

    unless (%opts) {
        $class->set_error('No options specified');
        return;
    }

    my $self = {
        keyserver_url => $opts{Keyserver} ? $opts{Keyserver} : $DEFAULT_KEYSERVER_URL,
        gnupg         => GnuPG::Interface->new(),
    };

    $self->{nodelist} = FTN::Crypt::Nodelist->new(
        Nodelist => $opts{Nodelist},
        Pointlist => $opts{Pointlist},
    );
    unless ($self->{nodelist}) {
        $class->set_error(FTN::Crypt::Nodelist->error);
        return;
    }

    $self->{gnupg}->options->hash_init(
        armor            => 1,
        meta_interactive => 0,
    );

    $self->{gnupg}->options->push_extra_args('--keyring', $opts{Pubring}) if $opts{Pubring};
    $self->{gnupg}->options->push_extra_args('--secret-keyring', $opts{Secring}) if $opts{Secring};
    $self->{gnupg}->options->push_extra_args('--always-trust');

    $self = bless $self, $class;
    return $self;
}

#----------------------------------------------------------------------#

=head2 encrypt_message()

Message encryption.

=head3 Parameters:

=over 4

=item * C<Address>: Recipient's FTN address.

=item * C<Message>: FTN message text with kludges.

=back

=head3 Returns:

Encrypted message or error in C<$obj-E<gt>error>.

Sample:

    my $res = $obj->encrypt_message(
        Address => $ftn_address,
        Message => $msg,
    ) or die $obj->error;

=cut

sub encrypt_message {
    my $self = shift;
    my (%opts) = @_;

    my $msg = FTN::Crypt::Msg->new(
        Address => $opts{Address},
        Message => $opts{Message},
    );
    unless ($msg) {
        $self->set_error(FTN::Crypt::Msg->error);
        return;
    }

    my $res;

    my ($addr, $method) = $self->{nodelist}->get_email_addr($msg->get_address);
    unless ($addr) {
        $self->set_error('Encryption-capable address not found', $self->{nodelist}->error);
        return;
    }

    my $gnupg_ver = $self->{gnupg}->version;
    if ($method eq 'PGP2') {
        if (version->parse($gnupg_ver) < version->parse($GPG2_BVER)) {
            $self->{gnupg}->options->meta_pgp_2_compatible(1);
        } else {
            $self->set_error("GnuPG is too new (ver. $gnupg_ver), can't ensure required encryption method ($method)");
            return;
        }
    } elsif ($method eq 'PGP5') {
        $self->{gnupg}->options->meta_pgp_5_compatible(1);
    }

    unless ($self->_lookup_key($addr) || $self->_import_key($addr)) {
        $self->set_error("PGP key for $addr not found");
        return;
    }
    
    my $key_id = $self->_select_key($addr);
    $self->{gnupg}->options->push_recipients($key_id);

    my ($in_fh, $out_fh, $err_fh) = (IO::Handle->new(), IO::Handle->new(),
        IO::Handle->new());

    my $handles = GnuPG::Handles->new(
        stdin  => $in_fh,
        stdout => $out_fh,
        stderr => $err_fh,
    );

    my $pid = $self->{gnupg}->encrypt(handles => $handles);

    print $in_fh $msg->get_text;
    close $in_fh;

    my $msg_enc = join '', <$out_fh>;
    close $out_fh;

    close $err_fh;
    
    waitpid $pid, 0;

    if ($msg_enc) {
        unless ($msg->set_text($msg_enc)) {
            $self->set_error("Can't write message text", $msg->error);
            return;
        }
        unless ($msg->add_kludge("$FTN::Crypt::Constants::ENC_MESSAGE_KLUDGE: $method")) {
            $self->set_error("Can't modify message kludges", $msg->error);
            return;
        }
        
        $res = $msg->get_message;
        unless ($res) {
            $self->set_error("Can't get message", $msg->error);
            return;
        }
    } else {
        $self->set_error('Message enccryption failed');
        return;
    }

    return $res;
}

#----------------------------------------------------------------------#

=head2 decrypt_message()

Message decryption.

=head3 Parameters:

=over 4

=item * C<Address>: Recipient's FTN address.

=item * C<Message>: FTN message text with kludges.

=item * C<Passphrase>: Key passphrase.

=back

=head3 Returns:

Decrypted message or error in C<$obj-E<gt>error>.

Sample:

    my $res = $obj->decrypt_message(
        Address => $ftn_address,
        Message => $msg,
        Passphrase => $pass,
    ) or die $obj->error;

=cut

sub decrypt_message {
    my $self = shift;
    my (%opts) = @_;

    unless (%opts) {
        $self->set_error('No options specified');
        return;
    }
    unless (defined $opts{Passphrase}) {
        $self->set_error('No passphrase specified');
        return;
    }

    my $msg = FTN::Crypt::Msg->new(
        Address => $opts{Address},
        Message => $opts{Message},
    );
    unless ($msg) {
        $self->set_error(FTN::Crypt::Msg->error);
        return;
    }

    my $res;

    my $method_used;
    foreach my $c (@{$msg->get_kludges}) {
        foreach my $k (@{$c}) {
            $method_used = $1 if $k =~ /^$FTN::Crypt::Constants::ENC_MESSAGE_KLUDGE:\s+(\w+)$/;
        }
    }
    unless ($method_used) {
        $self->set_error('Message seems not to be encrypted');
        return;
    }

    my ($addr, $method) = $self->{nodelist}->get_email_addr($msg->get_address);
    unless ($addr) {
        $self->set_error('Encryption-capable address not found', $self->{nodelist}->error);
        return;
    }
    unless ($method) {
        $self->set_error('Encryption method not found', $self->{nodelist}->error);
        return;
    }

    if ($method ne $method_used) {
        $self->set_error("Message is encrypted with $method_used while node uses $method");
        return;
    }

    my ($in_fh, $out_fh, $err_fh, $pass_fh) = (IO::Handle->new(),
        IO::Handle->new(), IO::Handle->new(), IO::Handle->new());
     
    my $handles = GnuPG::Handles->new(
        stdin      => $in_fh,
        stdout     => $out_fh,
        stderr     => $err_fh,
        passphrase => $pass_fh,
    );

    my $pid = $self->{gnupg}->decrypt(handles => $handles);

    print $pass_fh $opts{Passphrase};
    close $pass_fh;

    print $in_fh $msg->get_text;
    close $in_fh;
    
    my $msg_dec = join '', <$out_fh>;
    close $out_fh;

    close $err_fh;

    waitpid $pid, 0;

    if ($msg_dec) {
        unless ($msg->set_text($msg_dec)) {
            $self->set_error("Can't write message text", $msg->error);
            return;
        }
        unless ($msg->remove_kludge($FTN::Crypt::Constants::ENC_MESSAGE_KLUDGE)) {
            $self->set_error("Can't modify message kludges", $msg->error);
            return;
        }

        $res = $msg->get_message;
        unless ($res) {
            $self->set_error("Can't get message", $msg->error);
            return;
        }
    } else {
        $self->set_error('Message decryption failed');
        return;
    }

    return $res;
}

#----------------------------------------------------------------------#

sub _lookup_key {
    my $self = shift;
    my ($uid) = @_;
    
    my ($out_fh, $err_fh) = (IO::Handle->new(), IO::Handle->new());
     
    my $handles = GnuPG::Handles->new(
        stdout => $out_fh,
        stderr => $err_fh,
    );

    my $pid = $self->{gnupg}->list_public_keys(
        handles      => $handles,
        command_args => [$uid],
    );

    my $out = join '', <$out_fh>;

    close $out_fh;
    close $err_fh;

    waitpid $pid, 0;

    return $out ? 1 : 0;
}

#----------------------------------------------------------------------#

sub _import_key {
    my ($self, $uid) = @_;

    return if $self->_lookup_key($uid);

    my $res;

    try {
        my $finger = PGP::Finger->new(
            sources => [
                PGP::Finger::Keyserver->new(
                    url => $self->{keyserver_url},
                ),
                # Now there are no PGP-related DNS records in fidonet.net zone
                #~ PGP::Finger::DNS->new(
                    #~ dnssec => 1,
                    #~ rr_types => ['OPENPGPKEY', 'TYPE61'],
                #~ ),
            ],
        );
        $res = $finger->fetch($uid);
    } catch {
        return;
    };

    if ($res) {
        my ($in_fh, $out_fh, $err_fh) = (IO::Handle->new(),
             IO::Handle->new(), IO::Handle->new());
         
        my $handles = GnuPG::Handles->new(
            stdin  => $in_fh,
            stdout => $out_fh,
            stderr => $err_fh,
        );

        my $pid = $self->{gnupg}->import_keys(handles => $handles);

        print $in_fh $res->as_string('armored');
        close $in_fh;

        close $out_fh;
        close $err_fh;

        waitpid $pid, 0;

        return 1;
    }

    return;
}

#----------------------------------------------------------------------#

sub _select_key {
    my ($self, $uid) = @_;

    return unless $self->_lookup_key($uid);

    my @keys;
    foreach my $key ($self->{gnupg}->get_public_keys($uid)) {
        push @keys, [$key->creation_date, $key->hex_id]
            if !$self->_key_is_disabled($key) && $self->_key_can_encrypt($key);
        foreach my $subkey (@{$key->subkeys_ref}) {
            push @keys, [$subkey->creation_date, $subkey->hex_id]
                if !$self->_key_is_disabled($subkey) && $self->_key_can_encrypt($subkey);
        }
    }

    @keys = map { '0x' . substr $_->[1], -8 }
            sort { $b->[0] <=> $a->[0] }
            @keys;
    
    return $keys[0];
}

#----------------------------------------------------------------------#

sub _key_is_disabled {
    my ($self, $key) = @_;

    return index($key->usage_flags, 'D') != -1;
}

#----------------------------------------------------------------------#

sub _key_can_encrypt {
    my ($self, $key) = @_;

    return index($key->usage_flags, 'E') != -1;
}

1;
__END__

=head1 AUTHOR

Petr Antonov, E<lt>petr@antonov.spaceE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2019 by Petr Antonov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses at L<https://opensource.org/licenses/Artistic-1.0>, and
L<http://www.gnu.org/licenses/gpl-2.0.html>.

This package is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantability and fitness for a particular purpose.

=head1 INSTALLATION

Using C<cpan>:

    $ cpan FTN::Crypt

Manual install:

    $ perl Makefile.PL
    $ make
    $ make test
    $ make install

=head1 REFERENCES

=over 4

=item 1 L<FidoNet Policy Document Version 4.07|https://www.fidonet.org/policy4.txt>

=item 2 L<FTS-5001 - Nodelist flags and userflags|http://ftsc.org/docs/fts-5001.006>

=item 3 L<FSC-0073 - Encrypted message identification for FidoNet *Draft I*|http://ftsc.org/docs/fsc-0073.001>

=back 

=cut
